File: common/repository.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-

# Licensed under the GNU General Public License v. 3 (the "License")
# You may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.gnu.org/licenses/gpl-3.0.html
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Copyright (C) 2022-2024 xqyjlj<xqyjlj@126.com>
#
# @author      xqyjlj
# @file        repository.py
#
# Change Logs:
# Date           Author       Notes
# ------------   ----------   -----------------------------------------------
# 2024-10-07     xqyjlj       initial version
#

import json
import os

import jsonschema
import yaml
from loguru import logger

from .settings import SETTINGS


class RepositorySocType:
    class CurrentType:

        def __init__(self, data: dict):
            self.__data = data

        def __str__(self) -> str:
            return json.dumps(self.__data, indent=2, ensure_ascii=False)

        @property
        def origin(self) -> dict:
            return self.__data

        @property
        def lowest(self) -> float:
            return self.__data.get("lowest", 0)

        @property
        def run(self) -> float:
            return self.__data.get("run", 0)

    # ------------------------------------------------------------------------------------------------------------------
    class TemperatureType:

        def __init__(self, data: dict):
            self.__data = data

        def __str__(self) -> str:
            return json.dumps(self.__data, indent=2, ensure_ascii=False)

        @property
        def origin(self) -> dict:
            return self.__data

        @property
        def max(self) -> float:
            return self.__data.get("max", 0)

        @property
        def min(self) -> float:
            return self.__data.get("min", 0)

    # ------------------------------------------------------------------------------------------------------------------
    class VoltageType:

        def __init__(self, data: dict):
            self.__data = data

        def __str__(self) -> str:
            return json.dumps(self.__data, indent=2, ensure_ascii=False)

        @property
        def origin(self) -> dict:
            return self.__data

        @property
        def max(self) -> float:
            return self.__data.get("max", 0)

        @property
        def min(self) -> float:
            return self.__data.get("min", 0)

    # ------------------------------------------------------------------------------------------------------------------

    def __init__(
        self, data: dict, kind: str, vendor: str, series: str, line: str, name: str
    ):
        self.__data = data
        self.__current = None
        self.__temperature = None
        self.__voltage = None
        self.__kind = kind
        self.__vendor = vendor
        self.__series = series
        self.__line = line
        self.__name = name

    def __str__(self) -> str:
        return json.dumps(self.__data, indent=2, ensure_ascii=False)

    @property
    def kind(self) -> str:
        return self.__kind

    @property
    def vendor(self) -> str:
        return self.__vendor

    @property
    def series(self) -> str:
        return self.__series

    @property
    def line(self) -> str:
        return self.__line

    @property
    def name(self) -> str:
        return self.__name

    @property
    def origin(self) -> dict:
        return self.__data

    @property
    def core(self) -> str:
        return self.__data.get("core", "")

    @property
    def current(self) -> CurrentType:
        if self.__current is None:
            self.__current = RepositorySocType.CurrentType(
                self.__data.get("current", {})
            )
        return self.__current

    @property
    def flash(self) -> float:
        return self.__data.get("flash", 0)

    @property
    def frequency(self) -> float:
        return self.__data.get("frequency", 0)

    @property
    def io(self) -> int:
        return self.__data.get("io", 0)

    @property
    def package(self) -> str:
        return self.__data.get("package", "")

    @property
    def peripherals(self) -> dict[str, int]:
        return self.__data.get("peripherals", {})

    @property
    def ram(self) -> float:
        return self.__data.get("ram", 0)

    @property
    def temperature(self) -> TemperatureType:
        if self.__temperature is None:
            self.__temperature = RepositorySocType.TemperatureType(
                self.__data.get("temperature", {})
            )
        return self.__temperature

    @property
    def voltage(self) -> VoltageType:
        if self.__voltage is None:
            self.__voltage = RepositorySocType.VoltageType(
                self.__data.get("voltage", {})
            )
        return self.__voltage


class RepositoryType:
    def __init__(self, data: dict):
        self.__data = data
        self.__all_types = []
        self.__all_vendors = []
        self.__all_series = []
        self.__all_lines = []
        self.__all_soc = []
        self.__all_cores = []
        self.__all_package = []

        for kind, kind_item in self.__data.items():
            self.__all_types.append(kind)
            for vendor, vendor_item in kind_item.items():
                self.__all_vendors.append(vendor)
                for series, series_item in vendor_item.items():
                    self.__all_series.append(series)
                    for line, line_item in series_item.items():
                        self.__all_lines.append(line)
                        for soc, soc_item in line_item.items():
                            if kind == "soc":
                                item = RepositorySocType(
                                    soc_item, kind, vendor, series, line, soc
                                )
                                self.__all_soc.append(item)
                                self.__all_cores.append(item.core)
                                self.__all_package.append(item.package)

        self.__all_cores = list(set(self.__all_cores))
        self.__all_package = list(set(self.__all_package))

    def __str__(self) -> str:
        return json.dumps(self.__data, indent=2, ensure_ascii=False)

    @property
    def origin(self) -> dict[str, dict[str, dict[str, dict[str, dict[str, dict]]]]]:
        return self.__data

    def types(self) -> list[str]:
        return self.__all_types

    def all_types(self) -> list[str]:
        return self.types()

    def vendors(self, kind: str) -> list[str]:
        return list(self.__data.get(kind, {}).keys())

    def all_vendors(self) -> list[str]:
        return self.__all_vendors

    def series(self, kind: str, vendor: str) -> list[str]:
        return list(self.__data.get(kind, {}).get(vendor, {}).keys())

    def all_series(self) -> list[str]:
        return self.__all_series

    def lines(self, kind: str, vendor: str, series: str) -> list[str]:
        return list(self.__data.get(kind, {}).get(vendor, {}).get(series, {}).keys())

    def all_lines(self) -> list[str]:
        return self.__all_lines

    def names(self, kind: str, vendor: str, series: str, line: str) -> list[str]:
        return list(
            self.__data.get(kind, {})
            .get(vendor, {})
            .get(series, {})
            .get(line, {})
            .keys()
        )

    def __info(self, kind: str, vendor: str, series: str, line: str, name: str) -> dict:
        return (
            self.__data.get(kind, {})
            .get(vendor, {})
            .get(series, {})
            .get(line, {})
            .get(name, {})
        )

    def soc(
        self, kind: str, vendor: str, series: str, line: str, name: str
    ) -> RepositorySocType:
        return RepositorySocType(
            self.__info(kind, vendor, series, line, name),
            kind,
            vendor,
            series,
            line,
            name,
        )

    def all_soc(self) -> list[RepositorySocType]:
        return self.__all_soc

    def all_cores(self) -> list[str]:
        return self.__all_cores

    def all_package(self) -> list[str]:
        return self.__all_package


class Repository:

    def __init__(self):
        self.__repository = self.get_repository()

    @logger.catch(default=False)
    def __check_repository(self, repository: dict) -> bool:
        with open(
            os.path.join(SETTINGS.DATABASE_FOLDER, "schema", "repository.yml"),
            "r",
            encoding="utf-8",
        ) as f:
            schema = yaml.load(f.read(), Loader=yaml.FullLoader)
            jsonschema.validate(instance=repository, schema=schema)
        return True

    @logger.catch(default=RepositoryType({}))
    def __get_repository(self) -> RepositoryType:
        file = os.path.join(SETTINGS.DATABASE_FOLDER, "repository.yml")
        if os.path.isfile(file):
            with open(file, "r", encoding="utf-8") as f:
                repository = yaml.load(f.read(), Loader=yaml.FullLoader)
                succeed = self.__check_repository(repository)
            if succeed:
                return RepositoryType(repository)
            else:
                return RepositoryType({})
        else:
            logger.error(f"{file} is not file!")
            return RepositoryType({})

    def get_repository(self) -> RepositoryType:
        return self.__get_repository()

    def repository(self) -> RepositoryType:
        return self.__repository

File: src/database/repository.ts
/**
 * ****************************************************************************
 *  @author      xqyjlj
 *  @file        repository.ts
 *  @brief
 *
 * ****************************************************************************
 *  @attention
 *  Licensed under the Apache License v. 2 (the "License");
 *  You may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0.html
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  Copyright (C) 2025-2025 xqyjlj<xqyjlj@126.com>
 *
 * ****************************************************************************
 *  Change Logs:
 *  Date           Author       Notes
 *  ------------   ----------   -----------------------------------------------
 *  2025-08-14     xqyjlj       initial version
 */

import type {
  $Packages$,
  Hal,
  RepositoryChipSeriesType,
  RepositoryChipsLineType,
  RepositoryChipUnitType,
  RepositoryChipVendorType,
  RepositoryType,
} from '@/electron/types'
import type { App } from 'vue'
import { inject, markRaw } from 'vue'
import { I18n } from '~/i18n'

// #region typedef

export class Repository {
  private _origin: RepositoryType
  private _chips?: Record<string, RepositoryChipVendor>
  private _packages?: RepositoryPackages

  constructor(origin: RepositoryType) {
    this._origin = origin
  }

  get origin(): RepositoryType {
    return this._origin
  }

  get chips(): Record<string, RepositoryChipVendor> {
    if (!this._chips) {
      this._chips = {}
      const chipsData = this._origin.chips
      for (const [name, value] of Object.entries(chipsData)) {
        this._chips[name] = new RepositoryChipVendor(value)
      }
    }
    return this._chips
  }

  get packages(): RepositoryPackages {
    return this._packages ??= new RepositoryPackages(this._origin.packages)
  }

  get $packages$(): $Packages$ | undefined {
    return this._origin.$packages$
  }
}

export class RepositoryChipVendor {
  private _origin: RepositoryChipVendorType
  private _content?: Record<string, RepositoryChipSeries>

  constructor(_origin: RepositoryChipVendorType) {
    this._origin = _origin
  }

  get origin(): RepositoryChipVendorType {
    return this._origin
  }

  get content(): Record<string, RepositoryChipSeries> {
    if (!this._content) {
      this._content = {}
      const contentData = this._origin.content
      for (const [name, value] of Object.entries(contentData)) {
        this._content[name] = new RepositoryChipSeries(value)
      }
    }
    return this._content
  }
}

export class RepositoryChipSeries {
  private _origin: RepositoryChipSeriesType
  private _lines?: Record<string, RepositoryChipsLine>

  constructor(_origin: RepositoryChipSeriesType) {
    this._origin = _origin
  }

  get origin(): RepositoryChipSeriesType {
    return this._origin
  }

  get lines(): Record<string, RepositoryChipsLine> {
    if (!this._lines) {
      this._lines = {}
      for (const [name, value] of Object.entries(this._origin)) {
        this._lines[name] = new RepositoryChipsLine(value)
      }
    }
    return this._lines
  }
}

export class RepositoryChipsLine {
  private _origin: RepositoryChipsLineType
  private _units?: Record<string, RepositoryChipUnit>

  constructor(_origin: RepositoryChipsLineType) {
    this._origin = _origin
  }

  get origin(): RepositoryChipsLineType {
    return this._origin
  }

  get units(): Record<string, RepositoryChipUnit> {
    if (!this._units) {
      this._units = {}
      for (const [name, value] of Object.entries(this._origin)) {
        this._units[name] = new RepositoryChipUnit(value)
      }
    }
    return this._units
  }
}

export class RepositoryChipUnit {
  private _origin: RepositoryChipUnitType

  constructor(_origin: RepositoryChipUnitType) {
    this._origin = _origin
  }

  get origin(): RepositoryChipUnitType {
    return this._origin
  }

  get core(): string {
    return this._origin.core
  }

  get current(): { lowest: number, run: number } {
    return this._origin.current
  }

  get flash(): number {
    return this._origin.flash
  }

  get frequency(): number {
    return this._origin.frequency
  }

  get io(): number {
    return this._origin.io
  }

  get package(): string {
    return this._origin.package
  }

  get peripherals(): Record<string, number> {
    return this._origin.peripherals
  }

  get ram(): number {
    return this._origin.ram
  }

  get temperature(): { max: number, min: number } {
    return this._origin.temperature
  }

  get voltage(): { max: number, min: number } {
    return this._origin.voltage
  }
}

export class RepositoryPackages {
  private _origin: {
    hal: Hal
    toolchains: Hal
    components: {
      system?: Hal
    }
  }

  private _hal?: RepositoryHal
  private _toolchains?: RepositoryHal
  private _components?: RepositoryComponents

  constructor(_origin: {
    hal: Hal
    toolchains: Hal
    components: {
      system?: Hal
    }
  }) {
    this._origin = _origin
  }

  get origin(): {
    hal: Hal
    toolchains: Hal
    components: {
      system?: Hal
    }
  } {
    return this._origin
  }

  get hal(): RepositoryHal {
    return this._hal ??= new RepositoryHal(this._origin.hal)
  }

  get toolchains(): RepositoryHal {
    return this._toolchains ??= new RepositoryHal(this._origin.toolchains)
  }

  get components(): RepositoryComponents {
    return this._components ??= new RepositoryComponents(this._origin.components)
  }
}

export class RepositoryHal {
  private _origin: Hal
  private _packages?: Record<string, RepositoryPackage>

  constructor(_origin: Hal) {
    this._origin = _origin
  }

  get origin(): Hal {
    return this._origin
  }

  get packages(): Record<string, RepositoryPackage> {
    if (!this._packages) {
      this._packages = {}
      for (const [name, value] of Object.entries(this._origin)) {
        this._packages[name] = new RepositoryPackage(value)
      }
    }
    return this._packages
  }
}

export class RepositoryComponents {
  private _origin: {
    system?: Hal
  }

  private _system?: RepositoryHal

  constructor(_origin: {
    system?: Hal
  }) {
    this._origin = _origin
  }

  get origin(): {
    system?: Hal
  } {
    return this._origin
  }

  get system(): RepositoryHal | null {
    if (this._origin.system) {
      return this._system ??= new RepositoryHal(this._origin.system)
    }
    return null
  }
}

export class RepositoryPackage {
  private _origin: $Packages$
  private _description?: I18n
  private _url?: I18n

  constructor(_origin: $Packages$) {
    this._origin = _origin
  }

  get origin(): $Packages$ {
    return this._origin
  }

  get version(): Record<string, {
    urls: string[] | { windows?: string[], linux?: string[], [k: string]: unknown }
    note: { en: string, [k: string]: unknown }
    [k: string]: unknown
  }> | undefined {
    return this._origin.version
  }

  get author(): {
    name: string
    email: string
    website: { blog?: string, github?: string, [k: string]: unknown }
    [k: string]: unknown
  } | undefined {
    return this._origin.author
  }

  get license(): string | undefined {
    return this._origin.license
  }

  get vendor(): string | undefined {
    return this._origin.vendor
  }

  //   get vendorUrl(): Url | undefined {
  //     return this._origin.vendorUrl
  //   }

  get description(): I18n | null {
    if (this._origin.description) {
      return this._description ??= new I18n(this._origin.description)
    }
    return null
  }

  get url(): I18n | null {
    if (this._origin.url) {
      return this._url ??= new I18n(this._origin.url)
    }
    return null
  }

  get support(): string | undefined {
    return this._origin.support
  }
}

// #endregion

export class RepositoryManager {
  private _repository?: Repository

  private async _get(): Promise<Repository> {
    const content = await window.electron.invoke('database:getRepository') as RepositoryType
    return new Repository(content)
  }

  async get(): Promise<Repository> {
    return this._repository ??= markRaw(await this._get())
  }
}

export function createRepositoryManagerPlugin() {
  const manager = new RepositoryManager()

  return {
    manager,
    plugin: {
      install(app: App) {
        app.provide('database@repositoryManager', manager)
      },
    },
  }
}

export function useRepositoryManager(): RepositoryManager {
  return inject('database@repositoryManager')!
}
